mhs-reference-implementation/mhs/routing/tests/test_sds_handler.py

```diff
@@ -39,17 +41,75 @@
 
 class TestMHSAttributeLookupHandler(TestCase):
 
+    def setUp(self) -> None:
+        self.cache = dc.DictionaryCache()
+        self.mocked_client = mocks.mocked_sds_client()
+
     @async_test
     async def test_get_endpoint(self):
-        handler = sds.MHSAttributeLookupHandler(mocks.mocked_sds_client())
+        handler = sds.MHSAttributeLookupHandler(mocks.mocked_sds_client(), self.cache)
         attributes = await handler.retrieve_mhs_attributes(ODS_CODE, INTERACTION_ID)
         for key, value in expected_mhs_attributes.items():
             self.assertEqual(value, attributes[key])
 
-        # Assert exact number of attributes, minus the unique values
+        # Assert exact number of attributes
         self.assertEqual(len(attributes), len(expected_mhs_attributes))
 
     @async_test
     async def test_no_client(self):
         with self.assertRaises(ValueError):
-            sds.MHSAttributeLookupHandler(None)
+            sds.MHSAttributeLookupHandler(None, self.cache)
+
+    @async_test
+    async def test_no_cache(self):
+        with self.assertRaises(ValueError):
+            sds.MHSAttributeLookupHandler(mocks.mocked_sds_client(), None)
+
+    @async_test
+    async def test_value_added_to_cache(self):
+        handler = sds.MHSAttributeLookupHandler(mocks.mocked_sds_client(), self.cache)
+        handler.sds_client = mocks.mocked_sds_client()
+        future = asyncio.Future()
+        future.set_result(None)
+
+        handler.cache.add_cache_value = MagicMock()
```

Comment:
Why not just have a `MagicMock` cache, instead of setting all the parts of the cache to be `MagicMock`s. `MagicMock` recursively creates more mocks so you can do `MagicMock().this.iss.still.a.mock`

Reply:
By default `MagicMock` doesn't work with async and can't be awaited, so I have to specify the methods and that they return a future.